### ISSUE TITLE
kubeadm join is silently stuck when token is expired

### ISSUE THREAD

**Issue by non7top** [kubernetes/kubernetes#86936]
When trying to join
```
# kubeadm join kubes:6443 --token xxxxx     --discovery-token-ca-cert-hash sha256:xxxxxx
W0108 01:52:35.642353    2564 join.go:346] [preflight] WARNING: JoinControlPane.controlPlane settings will be ignored when control-plane flag is not set.
[preflight] Running pre-flight checks
```
and stuck

with v3 important details are revealed
```
I0108 01:52:49.430728    2700 token.go:188] [discovery] Trying to connect to API Server "kubes:6443"
I0108 01:52:49.431002    2700 token.go:73] [discovery] Created cluster-info discovery client, requesting info from "https://kubes:6443"
I0108 01:52:49.440595    2700 token.go:191] [discovery] Failed to connect to API Server "kubes:6443": token id "xxxx" is invalid for this cluster or it has expired. Use "kubeadm token create" on the control-plane node to create a new valid token
```

A request is to improve logging so that those token be shown always (probably change those to higher log level from INFO, warning or critical would be more appropriate probably. Anyway further actions are not possible and hanging without any logging is no good.

**Comment by neolit123** [comment]
/area kubeadm
/remove-kind feature
/priority awaiting-more-evidence

this is a known problem of versions older than v1.17.

we cannot backport the change to older versions like v1.16, so:
- make sure you have valid tokens
- use v1.17+

thanks.
/close

**Comment by non7top** [comment]
Your comment doesn't make it clear enough on whether v1.17 is affected or not, but I'm seeing this on v1.17.0.

I will assume I will need a version that is higher than v1.17.0

kubeadm version: &version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.0", GitCommit:"70132b0f130acc0bed193d9ba59dd186f0e634cf", GitTreeState:"clean", BuildDate:"2019-12-07T21:17:50Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"linux/amd64"}

**Comment by neolit123** [comment]
looks like the change is only in master, so this will land only in v1.18.0.
we cannot backport this to older versions as it's not a critical bug and the changes that fixed the bug were disruptive.

> kubeadm join is silently stuck when token is expired

duplicate of:
https://github.com/kubernetes/kubeadm/issues/1917

for < 1.18 you can control the discovery timeout in JoinConfiguration.

https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#Discovery
https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#pkg-constants

> A request is to improve logging so that those token be shown always (probably change those to higher log level from INFO, warning or critical would be more appropriate probably. Anyway further actions are not possible and hanging without any logging is no good.

fixed for 1.18.

**Comment by non7top** [comment]
Perfect, thank you for the details.
